ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos.hs

```diff
@@ -0,0 +1,364 @@
+  , tptsOCert :: OCert (TPraosDSIGN c) (TPraosKES c)
+  }
+  deriving Generic
+
+instance TPraosCrypto c => NoUnexpectedThunks (TPraosToSign c)
+deriving instance TPraosCrypto c => Show (TPraosToSign c)
+
+class ( HasHeader hdr
+      , SignedHeader hdr
+      , Cardano.Crypto.KES.Class.Signable (TPraosKES c) (Signed hdr)
+      ) => HeaderSupportsTPraos c hdr where
+
+  -- Because we are using the executable spec, rather than implementing the
+  -- protocol directly here, we have a fixed header type rather than an
+  -- abstraction. So we must introduce this method.
+  headerToBHeader
+    :: proxy (TPraos c)
+    -> hdr
+    -> BHeader (TPraosHash c) (TPraosDSIGN c) (TPraosKES c) (TPraosVRF c)
+
+forgeTPraosFields :: ( HasNodeState (TPraos c) m
+                    , MonadRandom m
+                    , TPraosCrypto c
+                    , Cardano.Crypto.KES.Class.Signable (TPraosKES c) toSign
+                    )
+                 => NodeConfig (TPraos c)
+                 -> TPraosProof c
+                 -> (TPraosToSign c -> toSign)
+                 -> m (TPraosFields c toSign)
+forgeTPraosFields TPraosNodeConfig{..}  TPraosProof{..} mkToSign = do
+    let icn@TPraosIsCoreNode{..} = tpraosIsCoreNode
+        (DiscVKey issuerVK) = ocertVkCold tpraosIsCoreNodeOpCert
+        signedFields = TPraosToSign {
+          tptsIssuerVK = issuerVK
+        , tptsVrfVk = deriveVerKeyVRF tpraosSignKeyVRF
+        , tptsEta = tpraosEta
+        , tptsLeader = tpraosLeader
+        , tptsOCert = tpraosIsCoreNodeOpCert
+        }
+    m <- signedKES
+           (fromIntegral (unSlotNo tpraosProofSlot))
+           (mkToSign signedFields)
+           tpraosIsCoreNodeSKSHot
+    case m of
+      Nothing                  -> error "mkOutoborosPayload: signedKES failed"
+      Just (signature, newKey) -> do
+        putNodeState . Just $ icn { tpraosIsCoreNodeSKSHot = newKey }
+        return $ TPraosFields {
+            tpraosSignature    = signature
+          , tpraosToSign       = signedFields
+          }
+
+{-------------------------------------------------------------------------------
+  TPraos specific types
+-------------------------------------------------------------------------------}
+
+-- | Assembled proof that the issuer has the right to issue a block in the
+-- selected slot.
+data TPraosProof c
+  = TPraosProof
+    { tpraosEta       :: CertifiedVRF (TPraosVRF c) Nonce
+    , tpraosLeader    :: CertifiedVRF (TPraosVRF c) UnitInterval
+    , tpraosProofSlot :: SlotNo
+    , tpraosIsCoreNode :: TPraosIsCoreNode c
+    } deriving Generic
+
+instance TPraosCrypto c => NoUnexpectedThunks (TPraosProof c)
+
+data TPraosLedgerView c = TPraosLedgerView {
+    -- | Stake distribution
+    tpraosLedgerViewPoolDistr :: PoolDistr (TPraosHash c) (TPraosDSIGN c) (TPraosVRF c)
+  , tpraosLedgerViewProtParams :: PParams
+  , tpraosLedgerViewDelegationMap :: Dms (TPraosHash c) (TPraosDSIGN c)
+  , tpraosLedgerViewEpochNonce :: Nonce
```

Comment:
Meaning, this won't be a concern for the hard fork combinator?

Reply:
It will move to being a pure protocol concern, so may still be a concern for the hard fork combinator. In particular, I believe that the spec calls for the insertion of extra entropy around the time of the hard fork.